test/mitmproxy/proxy2/layers/test_modes.py

```diff
@@ -0,0 +1,275 @@
+    """Test mitmproxy in reverse proxy mode.
+
+     - make sure that we connect to the right host
+     - make sure that we respect keep_host_header
+     - make sure that we include non-standard ports in the host header (#4280)
+    """
+    server = Placeholder(Server)
+    tctx.options.mode = "reverse:http://localhost:8000"
+    tctx.options.keep_host_header = keep_host_header
+    assert (
+            Playbook(modes.ReverseProxy(tctx), hooks=False)
+            >> DataReceived(tctx.client, b"GET /foo HTTP/1.1\r\n"
+                                         b"Host: example.com\r\n\r\n")
+            << NextLayerHook(Placeholder(NextLayer))
+            >> reply_next_layer(lambda ctx: http.HttpLayer(ctx, HTTPMode.transparent))
+            << OpenConnection(server)
+            >> reply(None)
+            << SendData(server, b"GET /foo HTTP/1.1\r\n"
+                                b"Host: " + (b"example.com" if keep_host_header else b"localhost:8000") + b"\r\n\r\n")
+            >> DataReceived(server, b"HTTP/1.1 200 OK\r\nContent-Length: 0\r\n\r\n")
+            << SendData(tctx.client, b"HTTP/1.1 200 OK\r\nContent-Length: 0\r\n\r\n")
+    )
+    assert server().address == ("localhost", 8000)
+
+
+@pytest.mark.parametrize("patch", [True, False])
+@pytest.mark.parametrize("connection_strategy", ["eager", "lazy"])
+def test_reverse_proxy_tcp_over_tls(tctx: Context, monkeypatch, patch, connection_strategy):
+    """
+    Test
+        client --TCP-- mitmproxy --TCP over TLS-- server
+    reverse proxying.
+    """
+
+    if patch:
+        monkeypatch.setattr(tls, "ServerTLSLayer", tls.MockTLSLayer)
+
+    flow = Placeholder(TCPFlow)
+    data = Placeholder(bytes)
+    tctx.options.mode = "reverse:https://localhost:8000"
+    tctx.options.connection_strategy = connection_strategy
+    playbook = Playbook(modes.ReverseProxy(tctx))
+    if connection_strategy == "eager":
+        (
+                playbook
+                << OpenConnection(tctx.server)
+                >> DataReceived(tctx.client, b"\x01\x02\x03")
+                >> reply(None, to=OpenConnection(tctx.server))
+        )
+    else:
+        (
+                playbook
+                >> DataReceived(tctx.client, b"\x01\x02\x03")
+        )
+    if patch:
+        (
+                playbook
+                << NextLayerHook(Placeholder(NextLayer))
+                >> reply_next_layer(tcp.TCPLayer)
+                << TcpStartHook(flow)
+                >> reply()
+        )
+        if connection_strategy == "lazy":
+            (
+                    playbook
+                    << OpenConnection(tctx.server)
+                    >> reply(None)
+            )
+        assert (
+                playbook
+                << TcpMessageHook(flow)
+                >> reply()
+                << SendData(tctx.server, data)
+        )
+        assert data() == b"\x01\x02\x03"
+    else:
+        if connection_strategy == "lazy":
+            (
```

Comment:
What is the difference between this block and the one below with assert? Is the assert statement not necessary?

Reply:
The assert statement starts the actual evaluation of the playbook. You can prepare a playbook for evaluation like this, but if you'd assert here you'd get an error: the final `>> reply(None)` triggers the `<< TlsStartHook(Placeholder())` below, which would not be covered here.
Unasserted playbooks automatically raise an error on deletion, so you can't forget to evaluate them. :)